File: app/src/main/java/com/osfans/trime/core/Rime.java
/*
 * Copyright (C) 2015-present, osfans
 * waxaca@163.com https://github.com/osfans
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.osfans.trime.core;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.osfans.trime.data.AppPrefs;
import com.osfans.trime.data.DataManager;
import com.osfans.trime.data.opencc.OpenCCDictManager;
import com.osfans.trime.data.schema.SchemaManager;
import java.util.Map;
import kotlin.Pair;
import kotlinx.coroutines.channels.BufferOverflow;
import kotlinx.coroutines.flow.FlowKt;
import kotlinx.coroutines.flow.MutableSharedFlow;
import kotlinx.coroutines.flow.SharedFlow;
import kotlinx.coroutines.flow.SharedFlowKt;
import timber.log.Timber;

/**
 * Rime與OpenCC的Java實現
 *
 * @see <a href="https://github.com/rime/librime">Rime</a> <a
 *     href="https://github.com/BYVoid/OpenCC">OpenCC</a>
 */
public class Rime {
  public final SharedFlow<RimeEvent> rimeNotiFlow = FlowKt.asSharedFlow(rimeNotiFlow_);

  /** Rime編碼區 */
  public static class RimeComposition {
    int length;
    int cursor_pos;
    int sel_start;
    int sel_end;
    String preedit;
    byte[] bytes;

    public String getText() {
      if (length == 0) return "";
      bytes = preedit.getBytes();
      return preedit;
    }

    public int getStart() {
      if (length == 0) return 0;
      return new String(bytes, 0, sel_start).length();
    }

    public int getEnd() {
      if (length == 0) return 0;
      return new String(bytes, 0, sel_end).length();
    }
  }

  /** Rime候選區，包含多個{@link CandidateListItem 候選項} */
  public static class RimeMenu {
    int page_size;
    int page_no;
    boolean is_last_page;
    int highlighted_candidate_index;
    int num_candidates;
    CandidateListItem[] candidates;
  }

  /** Rime上屏的字符串 */
  public static class RimeCommit {
    int data_size;
    // v0.9
    String text;
  }

  /** Rime環境，包括 {@link RimeComposition 編碼區} 、{@link RimeMenu 候選區} */
  public static class RimeContext {
    // v0.9
    RimeComposition composition;
    RimeMenu menu;
    // v0.9.2
    String commit_text_preview;
    String[] select_labels;

    public CandidateListItem[] getCandidates() {
      int numCandidates = menu != null ? menu.num_candidates : 0;
      Timber.d("setWindow getCandidates() numCandidates=%s", numCandidates);
      return numCandidates != 0 ? menu.candidates : new CandidateListItem[0];
    }
  }

  /** Rime狀態 */
  public static class RimeStatus {
    // v0.9
    String schema_id;
    String schema_name;
    boolean is_disabled;
    boolean is_composing;
    boolean is_ascii_mode;
    boolean is_full_shape;
    boolean is_simplified;
    boolean is_traditional;
    boolean is_ascii_punct;
  }

  private static Rime self;

  private static final RimeCommit mCommit = new RimeCommit();
  private static final RimeContext mContext = new RimeContext();
  private static final RimeStatus mStatus = new RimeStatus();
  private static boolean isHandlingRimeNotification;

  public static final MutableSharedFlow<RimeEvent> rimeNotiFlow_ =
      SharedFlowKt.MutableSharedFlow(0, 15, BufferOverflow.DROP_OLDEST);

  static {
    System.loadLibrary("rime_jni");
  }

  @NonNull
  private static AppPrefs getAppPrefs() {
    return AppPrefs.defaultInstance();
  }
  ;

  /*
  Android SDK包含了如下6个修饰键的状态，其中function键会被trime消费掉，因此只处理5个键
  Android和librime对按键命名并不一致。读取可能有误。librime按键命名见如下链接，
  https://github.com/rime/librime/blob/master/src/rime/key_table.cc
   */
  public static int META_SHIFT_ON = getRimeModifierByName("Shift");
  public static int META_CTRL_ON = getRimeModifierByName("Control");
  public static int META_ALT_ON = getRimeModifierByName("Alt");
  public static int META_SYM_ON = getRimeModifierByName("Super");
  public static int META_META_ON = getRimeModifierByName("Meta");

  public static int META_RELEASE_ON = getRimeModifierByName("Release");

  public static boolean hasMenu() {
    return isComposing() && mContext.menu.num_candidates != 0;
  }

  public static boolean hasLeft() {
    return hasMenu() && mContext.menu.page_no != 0;
  }

  public static boolean hasRight() {
    return hasMenu() && !mContext.menu.is_last_page;
  }

  public static boolean isPaging() {
    return hasLeft();
  }

  public static boolean isComposing() {
    return mStatus.is_composing;
  }

  public static boolean isAsciiMode() {
    return mStatus.is_ascii_mode;
  }

  public static boolean isAsciiPunch() {
    return mStatus.is_ascii_punct;
  }

  public static boolean showAsciiPunch() {
    return mStatus.is_ascii_punct || mStatus.is_ascii_mode;
  }

  public static RimeComposition getComposition() {
    if (mContext == null) return null;
    return mContext.composition;
  }

  public static String getCompositionText() {
    RimeComposition composition = getComposition();
    return (composition == null || composition.preedit == null) ? "" : composition.preedit;
  }

  public static String getComposingText() {
    if (mContext.commit_text_preview == null) return "";
    return mContext.commit_text_preview;
  }

  public Rime(boolean full_check) {
    startup(full_check);
    self = this;
  }

  public static void initSchema() {
    Timber.d("initSchema() RimeSchema");
    SchemaManager.init(getCurrentRimeSchema());
    Timber.d("initSchema() getStatus");
    getStatus();
  }

  @SuppressWarnings("UnusedReturnValue")
  private static boolean getStatus() {
    SchemaManager.updateSwitchOptions();
    return getRimeStatus(mStatus);
  }

  private static void startup(boolean full_check) {
    isHandlingRimeNotification = false;

    DataManager.sync();
    final String sharedDataDir = getAppPrefs().getProfile().getSharedDataDir();
    final String userDataDir = getAppPrefs().getProfile().getUserDataDir();

    Timber.i("Starting up Rime APIs ...");
    startupRime(sharedDataDir, userDataDir, full_check);

    Timber.i("Updating schema switchers ...");
    initSchema();
  }

  public static void deploy() {
    exitRime();
    startup(true);
  }

  public static String getCommitText() {
    return mCommit.text;
  }

  public static boolean getCommit() {
    return getRimeCommit(mCommit);
  }

  public static void getContexts() {
    Timber.i("\t<TrimeInput>\tgetContexts() get_context");
    // get_context() 是耗时操作
    getRimeContext(mContext);
    Timber.i("\t<TrimeInput>\tgetContexts() getStatus");
    getStatus();
    Timber.i("\t<TrimeInput>\tgetContexts() finish");
  }

  public static boolean isVoidKeycode(int keycode) {
    int XK_VoidSymbol = 0xffffff;
    return keycode <= 0 || keycode == XK_VoidSymbol;
  }

  // KeyProcess 调用JNI方法发送keycode和mask
  public static boolean processKey(int keycode, int mask) {
    Timber.i("\t<TrimeInput>\tonkey()\tkeycode=%s, mask=%s", keycode, mask);
    if (isVoidKeycode(keycode)) return false;
    // 此处调用native方法是耗时操作
    final boolean b = processRimeKey(keycode, mask);
    Timber.i(
        "\t<TrimeInput>\tonkey()\tkeycode=%s, mask=%s, process_key result=%s", keycode, mask, b);
    getContexts();
    Timber.i("\t<TrimeInput>\tonkey()\tfinish");
    return b;
  }

  public static boolean isValidText(CharSequence text) {
    if (text == null || text.length() == 0) return false;
    int ch = text.toString().codePointAt(0);
    return ch >= 0x20 && ch < 0x80;
  }

  public static boolean onText(CharSequence text) {
    if (!isValidText(text)) return false;
    boolean b = simulateKeySequence(text.toString().replace("{}", "{braceleft}{braceright}"));
    Timber.i("simulate key sequence = %s, input = %s", b, text);
    getContexts();
    return b;
  }

  public static CandidateListItem[] getCandidatesOrStatusSwitches() {
    final boolean showSwitches = getAppPrefs().getKeyboard().getSwitchesEnabled();
    if (!isComposing() && showSwitches) return SchemaManager.getStatusSwitches();
    return mContext.getCandidates();
  }

  public static CandidateListItem[] getCandidatesWithoutSwitch() {
    if (isComposing()) return mContext.getCandidates();
    return new CandidateListItem[0];
  }

  public static String[] getSelectLabels() {
    return mContext.select_labels;
  }

  public static int getCandHighlightIndex() {
    return isComposing() ? mContext.menu.highlighted_candidate_index : -1;
  }

  public static boolean commitComposition() {
    boolean b = commitRimeComposition();
    getContexts();
    return b;
  }

  public static void clearComposition() {
    clearRimeComposition();
    getContexts();
  }

  public static boolean selectCandidate(int index) {
    boolean b = selectRimeCandidateOnCurrentPage(index);
    getContexts();
    return b;
  }

  public static boolean deleteCandidate(int index) {
    boolean b = deleteRimeCandidateOnCurrentPage(index);
    getContexts();
    return b;
  }

  public static void setOption(String option, boolean value) {
    if (isHandlingRimeNotification) return;
    setRimeOption(option, value);
  }

  public static boolean getOption(String option) {
    return getRimeOption(option);
  }

  public static void toggleOption(String option) {
    boolean b = getOption(option);
    setOption(option, !b);
  }

  private static boolean isEmpty(@NonNull String s) {
    return s.contentEquals(".default"); // 無方案
  }

  public static boolean isEmpty() {
    return isEmpty(getCurrentRimeSchema());
  }

  public static String getSchemaName() {
    return mStatus.schema_name;
  }

  public static boolean selectSchema(String schemaId) {
    Timber.d("Selecting schemaId=%s", schemaId);
    boolean b = selectRimeSchema(schemaId);
    getContexts();
    return b;
  }

  public static Rime get(boolean full_check) {
    if (self == null) {
      if (full_check) {
        OpenCCDictManager.buildOpenCCDict();
      }
      self = new Rime(full_check);
    }
    return self;
  }

  public static Rime get() {
    return get(false);
  }

  public static void RimeSetCaretPos(int caret_pos) {
    setRimeCaretPos(caret_pos);
    getContexts();
  }

  public static void handleRimeNotification(
      @NonNull String messageType, @NonNull String messageValue) {
    isHandlingRimeNotification = true;
    final RimeEvent event = RimeEvent.create(messageType, messageValue);
    Timber.d("Handling Rime notification: %s", event);
    rimeNotiFlow_.tryEmit(event);
    isHandlingRimeNotification = false;
  }

  // init
  public static native void startupRime(
      @NonNull String sharedDir, @NonNull String userDir, boolean fullCheck);

  public static native void exitRime();

  public static native boolean deployRimeSchemaFile(@NonNull String schemaFile);

  public static native boolean deployRimeConfigFile(
      @NonNull String fileName, @NonNull String versionKey);

  public static native boolean syncRimeUserData();

  // input
  public static native boolean processRimeKey(int keycode, int mask);

  public static native boolean commitRimeComposition();

  public static native void clearRimeComposition();

  // output
  public static native boolean getRimeCommit(RimeCommit commit);

  public static native boolean getRimeContext(RimeContext context);

  public static native boolean getRimeStatus(RimeStatus status);

  // runtime options
  public static native void setRimeOption(@NonNull String option, boolean value);

  public static native boolean getRimeOption(@NonNull String option);

  @NonNull
  public static native SchemaListItem[] getRimeSchemaList();

  @NonNull
  public static native String getCurrentRimeSchema();

  public static native boolean selectRimeSchema(@NonNull String schemaId);

  @Nullable
  public static native Map<String, Object> getRimeConfigMap(
      @NonNull String configId, @NonNull String key);

  public static native void setRimeCustomConfigInt(
      @NonNull String configId, @NonNull Pair<String, Integer>[] keyValuePairs);

  // testing
  public static native boolean simulateKeySequence(@NonNull String keySequence);

  public static native String getRimeRawInput();

  public static native int getRimeCaretPos();

  public static native void setRimeCaretPos(int caretPos);

  public static native boolean selectRimeCandidateOnCurrentPage(int index);

  public static native boolean deleteRimeCandidateOnCurrentPage(int index);

  public static native String getLibrimeVersion();

  // module
  public static native boolean runRimeTask(String task_name);

  public static native String getRimeSharedDataDir();

  public static native String getRimeUserDataDir();

  public static native String getRimeSyncDir();

  public static native String getRimeUserId();

  // key_table
  public static native int getRimeModifierByName(@NonNull String name);

  public static native int getRimeKeycodeByName(@NonNull String name);

  @NonNull
  public static native SchemaListItem[] getAvailableRimeSchemaList();

  @NonNull
  public static native SchemaListItem[] getSelectedRimeSchemaList();

  public static native boolean selectRimeSchemas(@NonNull String[] schemaIds);

  @Nullable
  public static native String getRimeStateLabel(@NonNull String optionName, boolean state);
}

File: app/src/main/java/com/osfans/trime/core/Rime.kt
/*
 * SPDX-FileCopyrightText: 2015 - 2025 Rime community
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package com.osfans.trime.core

import com.osfans.trime.BuildConfig
import com.osfans.trime.data.base.DataManager
import com.osfans.trime.data.opencc.OpenCCDictManager
import com.osfans.trime.data.prefs.AppPrefs
import com.osfans.trime.ime.core.InlinePreeditMode
import com.osfans.trime.util.appContext
import com.osfans.trime.util.isStorageAvailable
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber

/**
 * Rime JNI and instance methods
 *
 * @see [librime](https://github.com/rime/librime)
 */
class Rime :
    RimeApi,
    RimeLifecycleOwner {
    private val lifecycleRegistry = RimeLifecycleRegistry()
    override val lifecycle get() = lifecycleRegistry

    override val messageFlow = messageFlow_.asSharedFlow()

    override val isReady: Boolean
        get() = lifecycle.currentState == RimeLifecycle.State.READY

    override var schemaCached = RimeSchema(".default")
        private set

    override var statusCached = StatusProto()
        private set

    override var compositionCached = CompositionProto()
        private set

    override var hasMenu: Boolean = false
        private set

    override var paging: Boolean = false
        private set

    private val dispatcher =
        RimeDispatcher(
            object : RimeDispatcher.RimeController {
                override fun nativeStartup() {
                    startRime(false)
                    lifecycleRegistry.emitState(RimeLifecycle.State.READY)
                }

                override fun nativeFinalize() {
                    exitRime()
                }
            },
        )

    private val inlinePreeditMode by AppPrefs.defaultInstance().general.inlinePreeditMode
    private val showAsciiSwitchTips by AppPrefs.defaultInstance().general.asciiSwitchTips
    private var lastAsciiTipsText = ""
    private var asciiSwitchTipsJob: Job? = null

    init {
        if (lifecycle.currentState != RimeLifecycle.State.STOPPED) {
            throw IllegalStateException("Rime has already been created!")
        }
    }

    private suspend inline fun <T> withRimeContext(crossinline block: suspend () -> T): T = withContext(dispatcher) {
        block()
    }

    override suspend fun isEmpty(): Boolean = withRimeContext {
        getCurrentRimeSchema() == ".default" // 無方案
    }

    override suspend fun deploy() = withRimeContext {
        exitRime()
        startRime(true)
    }

    override suspend fun syncUserData(): Boolean = withRimeContext {
        syncRimeUserData()
    }

    override suspend fun processKey(
        value: Int,
        modifiers: UInt,
        isVirtual: Boolean,
    ): Boolean = withRimeContext {
        processKeyInner(value, modifiers.toInt(), isVirtual)
    }

    override suspend fun processKey(
        value: KeyValue,
        modifiers: KeyModifiers,
        isVirtual: Boolean,
    ): Boolean = withRimeContext {
        processKeyInner(value.value, modifiers.toInt(), isVirtual)
    }

    override suspend fun simulateKeySequence(sequence: String): Boolean = withRimeContext {
        Timber.d("simulateKeySequence: $sequence")
        if (simulateRimeKeySequence(sequence)) {
            val commit = getRimeCommit()
            val input = getRimeRawInput()
            if (!commit.text.isNullOrEmpty() || input.isNotEmpty()) {
                emitResponse { commit }
                true
            } else {
                emitResponse { CommitProto(sequence) }
                false
            }
        } else {
            false
        }.also { Timber.d("simulateKeySequence ${if (it) "success" else "failed"}") }
    }

    override suspend fun selectCandidate(idx: Int, global: Boolean): Boolean = withRimeContext {
        selectRimeCandidate(idx, global).also { emitResponse() }
    }

    override suspend fun deleteCandidate(idx: Int, global: Boolean): Boolean = withRimeContext {
        deleteRimeCandidate(idx, global).also { emitResponse() }
    }

    override suspend fun changeCandidatePage(backward: Boolean): Boolean = withRimeContext {
        changeRimeCandidatePage(backward).also { emitResponse() }
    }

    override suspend fun moveCursorPos(position: Int) = withRimeContext {
        setRimeCaretPos(position)
        emitResponse()
    }

    override suspend fun availableSchemata(): Array<SchemaItem> = withRimeContext { getAvailableRimeSchemaList() }

    override suspend fun enabledSchemata(): Array<SchemaItem> = withRimeContext { getSelectedRimeSchemaList() }

    override suspend fun setEnabledSchemata(schemaIds: Array<String>) = withRimeContext { selectRimeSchemas(schemaIds) }

    override suspend fun selectedSchemata(): Array<SchemaItem> = withRimeContext { getRimeSchemaList() }

    override suspend fun selectedSchemaId(): String = withRimeContext { getCurrentRimeSchema() }

    override suspend fun selectSchema(schemaId: String) = withRimeContext { selectRimeSchema(schemaId) }

    override suspend fun currentSchema(): RimeSchema = withRimeContext {
        RimeSchema(getCurrentRimeSchema())
    }

    override suspend fun commitComposition(): Boolean = withRimeContext { commitRimeComposition().also { if (it) emitResponse() } }

    override suspend fun clearComposition() = withRimeContext {
        clearRimeComposition()
        emitResponse()
    }

    override suspend fun setRuntimeOption(
        option: String,
        value: Boolean,
    ): Unit = withRimeContext {
        setRimeOption(option, value)
    }

    override suspend fun getRuntimeOption(option: String): Boolean = withRimeContext {
        getRimeOption(option)
    }

    override suspend fun getCandidates(
        startIndex: Int,
        limit: Int,
    ): Array<CandidateItem> = withRimeContext {
        getRimeCandidates(startIndex, limit)
    }

    private fun startRime(fullCheck: Boolean) {
        DataManager.sync()
        val sharedDataDir = DataManager.sharedDataDir.absolutePath
        val userDataDir = DataManager.userDataDir.absolutePath
        Timber.d(
            """
            Starting rime with:
            sharedDataDir: $sharedDataDir
            userDataDir: $userDataDir
            fullCheck: $fullCheck
            """.trimIndent(),
        )
        startupRime(sharedDataDir, userDataDir, BuildConfig.BUILD_VERSION_NAME, fullCheck)
    }

    private fun processKeyInner(value: Int, modifiers: Int, isVirtual: Boolean): Boolean {
        lastAsciiTipsText = asciiTipsText
        val handled = processRimeKey(value, modifiers)
        emitResponse()
        if (!handled) {
            handleRimeMessage(
                10, // RimeMessage.MessageType.Key,
                arrayOf(value, modifiers, isVirtual),
            )
        }
        return handled
    }

    private val asciiTipsText: String
        get() {
            val status = getRimeStatus()
            return if (status.isAsciiMode) {
                "En"
            } else if (status.schemaName.isNotEmpty() &&
                !status.schemaName.startsWith('.')
            ) {
                status.schemaName.take(2)
            } else {
                ""
            }
        }

    private fun emitResponse(
        commit: (() -> CommitProto) = { getRimeCommit() },
    ) {
        handleRimeMessage(4, arrayOf(commit.invoke()))
        val context = getRimeContext()
        handlePreedit(context.composition)
        if (context.composition.length <= 0 && lastAsciiTipsText != asciiTipsText) {
            showAsciiSwitchTips()
        }
        if (getRimeOption("paging_mode")) {
            handleRimeMessage(7, arrayOf(context.menu))
        } else {
            val bulk = getRimeBulkCandidates()
            handleRimeMessage(9, bulk)
        }
        handleRimeMessage(8, arrayOf(getRimeStatus()))
    }

    private fun handlePreedit(composition: CompositionProto) {
        val mode = if (getRimeOption("no_inline_preedit")) {
            InlinePreeditMode.DISABLE
        } else {
            inlinePreeditMode
        }
        val inlinePreedit = when (mode) {
            InlinePreeditMode.DISABLE -> ""
            InlinePreeditMode.COMPOSING_TEXT -> composition.preedit ?: ""
            InlinePreeditMode.COMMIT_TEXT_PREVIEW -> composition.commitTextPreview ?: ""
        }
        val composition = if (mode == InlinePreeditMode.COMPOSING_TEXT) {
            CompositionProto()
        } else {
            composition
        }
        handleRimeMessage(5, arrayOf(inlinePreedit))
        handleRimeMessage(6, arrayOf(composition))
    }

    private fun handleRimeMessage(it: RimeMessage<*>) {
        when (it) {
            is RimeMessage.SchemaMessage -> {
                statusCached = getRimeStatus()
                schemaCached = RimeSchema(it.data.id)
            }
            is RimeMessage.OptionMessage -> {
                // Option change won't trigger response update
                val status = getRimeStatus()
                statusCached = status
                updateSchemaCached(status)
                if (it.data.option == "ascii_mode") {
                    showAsciiSwitchTips()
                }
            }
            is RimeMessage.DeployMessage -> {
                if (it.data == RimeMessage.DeployMessage.State.Start) {
                    OpenCCDictManager.buildOpenCCDict()
                }
            }
            is RimeMessage.CompositionMessage -> {
                val composition = it.data
                compositionCached = composition
            }
            is RimeMessage.CandidateMenuMessage -> {
                val menu = it.data
                paging = menu.pageNumber != 0
                hasMenu = menu.candidates.isNotEmpty()
            }
            is RimeMessage.CandidateListMessage -> {
                hasMenu = it.data.candidates.isNotEmpty()
            }
            is RimeMessage.StatusMessage -> {
                statusCached = it.data
                updateSchemaCached(it.data)
            }
            else -> {}
        }
    }

    private fun updateSchemaCached(status: StatusProto) {
        val (schemaId, schemaName) = status
        // Engine response update won't send SchemaMessage, but usually update RimeStatus
        if (schemaId != schemaCached.schemaId) {
            schemaCached = RimeSchema(schemaId)
            // notify downstream consumers that schema has changed
            messageFlow_.tryEmit(
                RimeMessage.SchemaMessage(
                    SchemaItem(schemaId, schemaName),
                ),
            )
        }
    }

    private fun showAsciiSwitchTips() {
        if (!showAsciiSwitchTips) return
        val tipsText = asciiTipsText
        if (tipsText.isEmpty()) return

        lastAsciiTipsText = tipsText

        val tips = CompositionProto(tipsText)
        messageFlow_.tryEmit(RimeMessage.CompositionMessage(tips))
        compositionCached = tips
        asciiSwitchTipsJob?.cancel()
        asciiSwitchTipsJob = lifecycleScope.launch {
            delay(1000L)
            val ctx = getRimeContext()
            handleRimeMessage(6, arrayOf(ctx.composition))
        }
    }

    fun startup() {
        if (!appContext.isStorageAvailable()) {
            Timber.w("Skip starting rime: storage not available!")
            return
        }
        if (lifecycle.currentState != RimeLifecycle.State.STOPPED) {
            Timber.w("Skip starting rime: not at stopped state!")
            return
        }
        registerRimeMessageHandler(::handleRimeMessage)
        lifecycleRegistry.emitState(RimeLifecycle.State.STARTING)
        dispatcher.start()
    }

    fun finalize() {
        if (lifecycle.currentState != RimeLifecycle.State.READY) {
            Timber.w("Skip stopping rime: not at ready state!")
            return
        }
        lifecycleRegistry.emitState(RimeLifecycle.State.STOPPING)
        Timber.i("Rime finalize()")
        dispatcher.stop().let {
            if (it.isNotEmpty()) {
                Timber.w("${it.size} job(s) didn't get a chance to run!")
            }
        }
        lifecycleRegistry.emitState(RimeLifecycle.State.STOPPED)
        unregisterRimeMessageHandler(::handleRimeMessage)
    }

    companion object {
        private val messageFlow_ =
            MutableSharedFlow<RimeMessage<*>>(
                extraBufferCapacity = 15,
                onBufferOverflow = BufferOverflow.DROP_OLDEST,
            )

        private val rimeMessageHandlers = ArrayList<(RimeMessage<*>) -> Unit>()

        init {
            System.loadLibrary("rime_jni")
        }

        // init
        @JvmStatic
        external fun startupRime(
            sharedDir: String,
            userDir: String,
            versionName: String,
            fullCheck: Boolean,
        )

        @JvmStatic
        external fun exitRime()

        @JvmStatic
        external fun deployRimeSchemaFile(schemaFile: String): Boolean

        @JvmStatic
        external fun deployRimeConfigFile(
            fileName: String,
            versionKey: String,
        ): Boolean

        @JvmStatic
        external fun syncRimeUserData(): Boolean

        // input
        @JvmStatic
        external fun processRimeKey(
            keycode: Int,
            mask: Int,
        ): Boolean

        @JvmStatic
        external fun commitRimeComposition(): Boolean

        @JvmStatic
        external fun clearRimeComposition()

        // output
        @JvmStatic
        external fun getRimeCommit(): CommitProto

        @JvmStatic
        external fun getRimeContext(): ContextProto

        @JvmStatic
        external fun getRimeStatus(): StatusProto

        // runtime options
        @JvmStatic
        external fun setRimeOption(
            option: String,
            value: Boolean,
        )

        @JvmStatic
        external fun getRimeOption(option: String): Boolean

        @JvmStatic
        external fun getRimeSchemaList(): Array<SchemaItem>

        @JvmStatic
        external fun getCurrentRimeSchema(): String

        @JvmStatic
        external fun selectRimeSchema(schemaId: String): Boolean

        // testing
        @JvmStatic
        external fun simulateRimeKeySequence(keySequence: String): Boolean

        @JvmStatic
        external fun getRimeRawInput(): String

        @JvmStatic
        external fun getRimeCaretPos(): Int

        @JvmStatic
        external fun setRimeCaretPos(caretPos: Int)

        @JvmStatic
        external fun selectRimeCandidate(index: Int, global: Boolean): Boolean

        @JvmStatic
        external fun deleteRimeCandidate(index: Int, global: Boolean): Boolean

        @JvmStatic
        external fun changeRimeCandidatePage(backward: Boolean): Boolean

        @JvmStatic
        external fun getAvailableRimeSchemaList(): Array<SchemaItem>

        @JvmStatic
        external fun getSelectedRimeSchemaList(): Array<SchemaItem>

        @JvmStatic
        external fun selectRimeSchemas(schemaIds: Array<String>): Boolean

        @JvmStatic
        external fun getRimeCandidates(
            startIndex: Int,
            limit: Int,
        ): Array<CandidateItem>

        @JvmStatic
        external fun getRimeBulkCandidates(): Array<Any>

        @JvmStatic
        fun handleRimeMessage(
            type: Int,
            params: Array<Any>,
        ) {
            val message = RimeMessage.nativeCreate(type, params)
            Timber.d("Handling $message")
            rimeMessageHandlers.forEach { it.invoke(message) }
            messageFlow_.tryEmit(message)
        }

        private fun registerRimeMessageHandler(handler: (RimeMessage<*>) -> Unit) {
            if (rimeMessageHandlers.contains(handler)) return
            rimeMessageHandlers.add(handler)
        }

        private fun unregisterRimeMessageHandler(handler: (RimeMessage<*>) -> Unit) {
            rimeMessageHandlers.remove(handler)
        }
    }
}
